File: SPOJ/BYTESM2.cpp
#include <iostream>

using namespace std;

int matrix[101][101] = { 0 };
int M;
int N;

void buildMatrix();
int findMaxPath();
int max(int,int);

int main()
{
	int tests;
	cin >> tests;
	
	for (int i = 0; i < tests; ++i) {
		cin >> M >> N;
		
		buildMatrix();
		
		int max = findMaxPath();
		cout << max << endl;
	}
	
	return 0;
}

void buildMatrix() {
	for (int r = 0; r < M; ++r) 
		for (int c = 0; c < N; ++c)
			cin >> matrix[r][c];
}

int findMaxPath() {
	int temp = 0;
	for (int r = M-2; r >= 0; --r) {
		for (int c = 0; c < N; ++c) {
			for (int i = c-1; i <= c+1; ++i) {
				if (i > -1 && i < N)
					temp = max(temp,matrix[r+1][i]);
			}
			matrix[r][c] += temp;
			temp = 0;
		}
	}
	
	int pathMax = 0;
	for (int k = 0; k < N; ++k) 
		pathMax = ((matrix[0][k] > pathMax) ? matrix[0][k] : pathMax);
		
	return pathMax;
}

int max(int a, int b) {
	return ((a > b) ? a : b);
}

File: SPOJ/BYTESM2.java
import java.util.*;

class Main {
	private static Scanner in = new Scanner(System.in);
	private static int[][] matrix;
	private static int M;
	private static int N;
	public static void main(String[] args) {
		int tests = in.nextInt();
		
		for (int i = 0; i < tests; ++i) {
			matrix = buildMatrix();
			
			int most = findMaxPath();
			System.out.printf("%d\n", most);
		}
	}
	
	/** BUILD THE MATRIX FOR THE CURRENT TEST **/
	public static int[][] buildMatrix() {
		M = in.nextInt();
		N = in.nextInt();
		
		int[][] temp = new int[M][N];
		
		for (int r = 0; r < M; ++r) {
			for (int c = 0; c < N; ++c) {
				temp[r][c] = in.nextInt();
			}
		}
		
		return temp;
	}
	
	/** FIND THE MAXIMUM PATH USING RECURSION **/	
	public static int findMaxPath() {
		int temp = 0;
		for (int r = M-2; r >= 0; --r) {
			for (int c = 0; c < N; ++c) {
				for (int i = c-1; i <= c+1; ++i) {
					if (i > -1 && i < N)
						temp = max(temp,matrix[r+1][i]);
				}
				matrix[r][c] += temp;
				temp = 0;
			}
		}
		
		int pathMax = 0;
		for (int y = 0; y < N; ++y)
			pathMax = ((matrix[0][y] > pathMax) ? pathMax = matrix[0][y] : pathMax);
			
		return pathMax;
	}
	
	public static int max(int a, int b) {
		return ((a > b) ? a : b);
	}
}
